modules/migrate.py

```diff
@@ -0,0 +1,485 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A script to migrate from old style widgets & patterns to modular widgets and
+patterns.
+
+This script will check out the entire trunk of a project repo. It must do this
+to ensure it correctly fixes all relative paths referenced from a widget. This
+can be slow! If the repo is already checked out you can substitute the path
+to the repo for the repo shortname and the existing svn repository will be
+re-used and the environment will be ignored.
+
+This script uses a series of very targetted, programatically generated regular
+expressions to do the actual content migration. As a result the diff is very
+specific to the changes that are required and easy to review. If the content
+was parsed and re-rendered to HTML almost every line would change to do some
+formatting difference between python's libraries and Habitat's stack. While
+extensively tested, bugs happen, especially in complex regular expressions.
+
+REVIEW ALL PROJECTS MIGRATED BY THIS SCRIPT FOR CORRECTNESS!
+
+Configuration files must be a CSV in the form of:
+
+repo_shortname,environment,widget_directory_name,module_directory_name
+
+where the widget directory name is relative to /assets/widgets and the module
+directory name is relative to /assets/modules, e.g.,
+
+sn_test_project,stable,flashcard,inkling.flashcard
+
+Example command lines:
+    migrate.py -c migrate.csv
+    migrate.py -c migrate.csv -s
+"""
+
+import argparse
+import csv
+import glob
+import json
+import logging
+import os
+import re
+import subprocess
+
+from bs4 import BeautifulSoup
+
+from s9logging import s9logging
+import svn.project_svn as svn
+
+parser = argparse.ArgumentParser(description='Migrate content from a widget to '
+    'modularized form of the same widget in a given project.')
+parser.add_argument('-c', '--config', required=True, help='CSV file specifying '
+    'widgets migrating to modules in each project, each row in the form "'
+    'repo_shortname,environment,widget_directory_name,module_directory_name"')
+parser.add_argument('-s', '--skip-commit', action='store_true', default=False,
+    help='Whether to skip svn commit of project changes')
+
+s9logging.configureLogging()
+log = logging.getLogger(__name__)
+
+def _getSpecsFromCsv():
+    """Returns a list of tuples of the form (source name, source environment,
+    widget directory name, module directory name) taken from the CSV
+    configuration file.
+    """
+    results = []
+    if args.config:
+        with open(args.config, 'rb') as f:
+            reader = csv.reader(f)
+            try:
+                for row in reader:
+                    # Strip any whitespace in row contents.
+                    row = map(str.strip, row)
```

Comment:
nit: It is generally more pythonic to use list comprehension instead of map. Something like `row = [column.strip() for column in row]` would also help get the message across that row is a list of column values. For a sec, I thought row was 'col1,col2,col3'.